### DISCUSSION TITLE
How to scale azure function instances on premise?

### DISCUSSION THREAD

**Discussion by Insadem** [Azure/azure-functions-durable-extension#2481] [Q&A]
Hello, I deployed azure function app on vm, it works great.
I use azure storage as data storage provider and it works great too.

However, when I add another vm there's concurrency which leads to problems (such as processing data 2 times). 
As far as I understand azure cloud scales azure app across multiple instances, but why I can't do the same on premises?

It's 100% not related to my code implementation, because I already tried to run it on azure cloud and it worked great.
My host.json:
`{
  "version": "2.0",
  "logging": {
    "fileLoggingMode": "debugOnly",
    "applicationInsights": {
      "samplingSettings": {
        "isEnabled": false,
        "initialSamplingPercentage": 0.0,
        "minSamplingPercentage": 0.0,
        "maxSamplingPercentage": 0.0
      },
      "enableLiveMetrics": false,
      "enableDependencyTracking": false,
      "enablePerformanceCountersCollection": false,
      "httpAutoCollectionOptions": {
        "enableHttpTriggerExtendedInfoCollection": false,
        "enableW3CDistributedTracing": false,
        "enableResponseHeaderInjection": false
      }
    }
  },
  "concurrency": {
    "dynamicConcurrencyEnabled": true,
    "snapshotPersistenceEnabled": true
  },
  "extensions": {
    "durableTask": {
      "hubName": "DurableFunctionsHub",
      "storageProvider": {
        "connectionStringName": "AzureWebJobsStorage",
        "trackingStoreNamePrefix": "DurableTask",
        "controlQueueBatchSize": 32,
        "controlQueueBufferThreshold": 256,
        "maxQueuePollingInterval": "00:00:01"
      },
      "tracing": {
        "traceInputsAndOutputs": false,
        "traceReplayEvents": false
      },
      "maxConcurrentActivityFunctions": 30,
      "maxConcurrentOrchestratorFunctions": 50,
      "maxEntityOperationBatchSize": 1000
    }
  }
}
`

**Comment by vigouredelaruse** [comment]
in case i have to answer this queston for points on the package i looked into it
* reviewed https://learn.microsoft.com/en-us/azure/azure-functions/functions-concurrency and can see that the concurrency settings matter in host.json
* reviewed scale controlle activity https://learn.microsoft.com/en-us/azure/azure-functions/event-driven-scaling?tabs=azure-cli
* reviewed keda documentation and see the claim is seamless autoscaling https://learn.microsoft.com/en-us/azure/azure-functions/functions-kubernetes-keda

note that the only on-prem configurations discussed here (https://learn.microsoft.com/en-us/azure/architecture/hybrid/azure-functions-hybrid) refer to docker/kubernetes/keda

having said that, i have yet to encounter concurrency issues on prem in a kubernetes/keda configuration, with rabbitmq triggers though
